internal/datastore/memdb/schema.go

```diff
@@ -76,7 +97,11 @@ func (r relationship) Relationship() *v1.Relationship {
 	}
 }
 
-func (r relationship) RelationTuple() *core.RelationTuple {
+func (r relationship) RelationTuple() (*core.RelationTuple, error) {
+	cr, err := r.caveat.CaveatReference()
```

Comment:
I wonder if it would be better to "parse" the caveat at the caller site; most relationships loaded won't necessarily need the fully struct-ed context, and constructing it here is going to allocate a lot more times that will be necessary. Instead, maybe we just RelationTuple store the raw form, and add a helper to retrieve it (which can return an error at *that* time)?

Reply:
I'm not sure I fully grasp the idea. This method is used in a handful of places where it would seem like the exact representation of the `core.RelationTuple` is needed.
- It's used to [insert into the changelog table](https://github.com/vroldanbet/spicedb/blob/dff6f68808862f4fca561c3f3491921b6d763cac/internal/datastore/memdb/memdb.go#L197-L216) that powers `Watch` API, and hence we need the context returned back to the client. It's true that we could store the memdb-internal struct instead of the proto and avoid the internal-to-core conversion, but for some reason memdb is storing full blown proto values for the `changelog` but not for the `relationshipTuple` table. I could address that in a follow up, but seems a problem that was already in place before this PR
- it's used to implement the `RelationshipIterator`, which uses `core.RelationTuple`. Once again we would be returning the full Tuple representation in order to send it back to the client
- It's used in the `WriteRelationship` [path](https://github.com/vroldanbet/spicedb/blob/77b11ff10f71381c8db27bb1071b197a0ac00988/internal/datastore/memdb/readwrite.go#L70) to return a human readable error on duplicates. Not sure it's worth optimizing the error path.
- It's used to to generate the mutations needed [when a namespace is deleted](https://github.com/vroldanbet/spicedb/blob/77b11ff10f71381c8db27bb1071b197a0ac00988/internal/datastore/memdb/readwrite.go#L130). This is certainly an inefficient conversion given the system is [about to turns those again into their memdb-internal representation](https://github.com/vroldanbet/spicedb/blob/77b11ff10f71381c8db27bb1071b197a0ac00988/internal/datastore/memdb/readwrite.go#L38-L46) to insert then. This is a relatively easy refactor if you feel strongly about it. It also led me realize that we are not inserting in the changelog those delete tuples caused by a namespace deletion? Seems like a bug?

Overall it wasn't clear to me if optimization is a focus on the memdb implementation. I was under the impression it's meant to serve as reference implementation + testing environments. But happy to do any potential adjustments wherever it makes sense!

Reply:
Its more the need to raise an error in the iterator, but I guess that's fine